### PR TITLE
Use pydantic to encapsulate dataset & dataset search param info

### PR DESCRIPTION
**Pull request by trey-stafford**
This will allow custom validators against e.g,. version strings and short names. It also makes it possible to use the specific instance of `Dataset` singledispatch readers

### CODE REVIEW

File: src/iceflow/data/models.py [hunk 1]
```diff
@@ -44,3 +46,20 @@ class ATM1BSchema(CommonDataColumnsSchema):
...
+
+
+class ATM1BDataset(Dataset):
+    short_name: DatasetShortName = "ILATM1B"
+
+
+class DatasetSearchParameters(pydantic.BaseModel):
+    dataset: Dataset
+    bounding_box: Sequence[float]
+    temporal: tuple[str, str]
```
Comment (mfisher87) [inline src/iceflow/data/models.py:65]: What about `tuple[datetime, datetime]`? The type can be coerced from string, I think by default?

File: src/iceflow/data/read.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+@functools.singledispatch
+def read_data(dataset: Dataset, _filepath: Path) -> IceflowDataFrame:
+    msg = f"{dataset=} not recognized."
+    raise RuntimeError(msg)
+
+
+@read_data.register
+def _(_dataset: ATM1BDataset, filepath: Path) -> ATM1BDataFrame:
+    return atm1b_data(filepath)
```
Comment (mfisher87) [inline src/iceflow/data/read.py:23]: Sick

**Review by mfisher87 — APPROVED** [review]

File: src/iceflow/data/models.py (hunk 1 above)
Comment (trey-stafford) [reply src/iceflow/data/models.py:65]: Good suggestion!

Comment (trey-stafford) [reply src/iceflow/data/models.py]: e9fd548 * origin/pydantic-for-datasets Type temporal range as tuple[date, date]

I typed this as `tuple[dt.datetime | dt.date, dt.datetime | dt.date]`, as this is passed straight to `earthaccess` and it seems to support both. I like being able to just specify a date range without times, but users may want to specify time ranges too. 

Do you have thoughts on if I should just make this accept datetimes only vs dates & datetimes? I

File: src/iceflow/data/fetch.py [hunk 3]
```diff
@@ -11,7 +12,7 @@ def search_and_download(
     version: str,
     short_name: str,
     bounding_box: Sequence[float],
-    temporal: tuple[str, str],
+    temporal: tuple[dt.datetime | dt.date, dt.datetime | dt.date],
```
Comment (mfisher87) [inline src/iceflow/data/fetch.py:15]: You can technically do `tuple[dt.date, dt.date]`, as `dt.datetimes` are `dt.dates` too:

```python
>>> import datetime as dt
>>> date = dt.datetime.today().date()
>>> isinstance(date, dt.datetime)
False
>>> isinstance(date, dt.date)
True
>>> datetime = dt.datetime.today()
>>> isinstance(datetime, dt.datetime)
True
>>> isinstance(datetime, dt.date)
True
```

**Pull request merged**